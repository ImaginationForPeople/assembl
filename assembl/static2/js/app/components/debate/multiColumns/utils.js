// @flow

export const orderPostsByMessageClassifier = (messageColumns: IdeaMessageColumns, posts: Array<Post>) =>
  messageColumns.reduce((naziLinter, col) => {
    const keyName = col.messageClassifier;
    const columnsMap = naziLinter;
    columnsMap[keyName] = posts.filter(post => post.messageClassifier === keyName);
    return columnsMap;
  }, {});