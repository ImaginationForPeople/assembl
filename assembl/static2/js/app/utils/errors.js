/*
  All special errors are placed here. Creating special errors is very useful
  to manage results of async Action Creators, amongst other things.
*/

export class PasswordMismatchError extends Error {
  constructor(...args) {
    super(...args);
    this.name = 'PasswordMismatchError';
  }
}