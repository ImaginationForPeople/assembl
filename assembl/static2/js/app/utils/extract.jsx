// @flow
import React from 'react';
import get from 'lodash/get';
import ARange from 'annotator_range'; // eslint-disable-line

import { type ColorDefinition, type ExtractState, ExtractStates, harvestingColors, harvestingColorsMapping } from '../constants';
import DesignFiction from '../components/svg/taxonomy/displayDesignFiction';
import MultiColumn from '../components/svg/taxonomy/displayMultiColumn';
import OpenQuestions from '../components/svg/taxonomy/displayOpenQuestions';
import Thread from '../components/svg/taxonomy/displayThread';
import Tokens from '../components/svg/taxonomy/displayTokens';
import ToArgument from '../components/svg/taxonomy/toArgument';
import MixMatch from '../components/svg/taxonomy/mixMatch';
import Example from '../components/svg/taxonomy/example';
import Classify from '../components/svg/taxonomy/classify';
import MoreSpecific from '../components/svg/taxonomy/moreSpecific';
import MakeGeneric from '../components/svg/taxonomy/makeGeneric';
import Flag from '../components/svg/taxonomy/Flag';

export type Annotation = {
  body: string,
  xpathStart: string,
  xpathEnd: string,
  offsetStart: string,
  offsetEnd: string
};

export const extractNatures = [
  {
    qualifier: 'issue',
    label: 'search.taxonomy_nature.issue'
  },
  {
    qualifier: 'actionable_solution',
    label: 'search.taxonomy_nature.actionable_solution'
  },
  {
    qualifier: 'knowledge',
    label: 'search.taxonomy_nature.knowledge'
  },
  {
    qualifier: 'example',
    label: 'search.taxonomy_nature.example'
  },
  {
    qualifier: 'concept',
    label: 'search.taxonomy_nature.concept'
  },
  {
    qualifier: 'argument',
    label: 'search.taxonomy_nature.argument'
  },
  {
    qualifier: 'cognitive_bias',
    label: 'search.taxonomy_nature.cognitive_bias'
  }
];

export const extractActions = [
  {
    qualifier: 'classify',
    label: 'search.taxonomy_action.classify'
  },
  {
    qualifier: 'make_generic',
    label: 'search.taxonomy_action.make_generic'
  },
  {
    qualifier: 'argument',
    label: 'search.taxonomy_action.argument'
  },
  {
    qualifier: 'give_examples',
    label: 'search.taxonomy_action.give_examples'
  },
  {
    qualifier: 'more_specific',
    label: 'search.taxonomy_action.more_specific'
  },
  {
    qualifier: 'mix_match',
    label: 'search.taxonomy_action.mix_match'
  },
  {
    qualifier: 'display_multi_column',
    label: 'search.taxonomy_action.display_multi_column'
  },
  {
    qualifier: 'display_thread',
    label: 'search.taxonomy_action.display_thread'
  },
  {
    qualifier: 'display_tokens',
    label: 'search.taxonomy_action.display_tokens'
  },
  {
    qualifier: 'display_open_questions',
    label: 'search.taxonomy_action.display_open_questions'
  },
  {
    qualifier: 'display_bright_mirror',
    label: 'search.taxonomy_action.display_bright_mirror'
  }
];

type ActionIconsProps = {
  qualifier: string,
  backgroundColor: string,
  color: string
};

export const ActionIcons = ({ qualifier, backgroundColor, color }: ActionIconsProps) => {
  switch (qualifier) {
  case 'display_bright_mirror':
    return <DesignFiction backgroundColor={backgroundColor} color={color} />;
  case 'display_multi_column':
    return <MultiColumn backgroundColor={backgroundColor} color={color} />;
  case 'display_open_questions':
    return <OpenQuestions backgroundColor={backgroundColor} color={color} />;
  case 'display_thread':
    return <Thread backgroundColor={backgroundColor} color={color} />;
  case 'display_tokens':
    return <Tokens backgroundColor={backgroundColor} color={color} />;
  case 'argument':
    return <ToArgument backgroundColor={backgroundColor} color={color} />;
  case 'mix_match':
    return <MixMatch backgroundColor={backgroundColor} color={color} />;
  case 'give_examples':
    return <Example backgroundColor={backgroundColor} color={color} />;
  case 'classify':
    return <Classify backgroundColor={backgroundColor} color={color} />;
  case 'more_specific':
    return <MoreSpecific backgroundColor={backgroundColor} color={color} />;
  case 'make_generic':
    return <MakeGeneric backgroundColor={backgroundColor} color={color} />;
  default:
    return <span>{qualifier}</span>;
  }
};

function getTaxonomyNatureColor(nature: string): ColorDefinition | null {
  return get(harvestingColorsMapping, nature, null);
}

type NatureIconProps = {
  qualifier: string
};

export const NatureIcons = ({ qualifier }: NatureIconProps) => {
  const color = getTaxonomyNatureColor(qualifier);
  if (color) {
    return <Flag color={color.background} />;
  }

  return <span>{qualifier}</span>;
};

export const getExtractTagId = (id: number) => `message-body-local:Content/${id}`;

export function getExtractColor(nature: string, state: ExtractState, extractedByMachine: boolean): ColorDefinition {
  const defaultColor = harvestingColors.green2;
  if (extractedByMachine) {
    if (state === ExtractStates.SUBMITTED) {
      return harvestingColors.pink;
    }
    if (state === ExtractStates.PUBLISHED) {
      return harvestingColors.paleGreen;
    }
  }

  const natureColor = getTaxonomyNatureColor(nature.replace('Enum.', ''));
  return natureColor || defaultColor;
}

export const getAnnotationData = (selection: Object): Annotation | null => {
  const body = selection ? selection.toString() : '';
  const annotatorRange = selection && body ? ARange.sniff(selection.getRangeAt(0)) : null;
  const serializedAnnotatorRange = annotatorRange ? annotatorRange.serialize(document, 'annotation') : null;
  if (serializedAnnotatorRange) {
    return {
      body: body,
      xpathStart: serializedAnnotatorRange.start,
      xpathEnd: serializedAnnotatorRange.end,
      offsetStart: serializedAnnotatorRange.startOffset,
      offsetEnd: serializedAnnotatorRange.endOffset
    };
  }
  return null;
};