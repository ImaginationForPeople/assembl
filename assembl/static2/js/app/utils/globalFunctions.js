// @flow

const getInputValue = (id: string) => {
  const elem = document.getElementById(id);
  const value = elem instanceof HTMLInputElement ? elem.value : null;
  return value;
};

const getScriptText = (id: string) => {
  const elem = document.getElementById(id);
  const text = elem instanceof HTMLScriptElement ? elem.text : null;
  return text;
};

export const getDiscussionId = () => getInputValue('discussion-id');

export const getDiscussionSlug = () => getInputValue('discussion-slug');

// cache userId to avoid accessing the dom at each permission check
let userId;
export const getConnectedUserId = () => {
  if (userId === undefined) {
    userId = getInputValue('user-id');
  }
  return userId;
};

export const getConnectedUserName = () => getInputValue('user-displayname');

// cache permissions to avoid accessing the dom at each permission check
let permissions;
export const getConnectedUserPermissions = () => {
  if (permissions === undefined) {
    const scriptText = getScriptText('permissions-json') || '[]';
    permissions = JSON.parse(scriptText);
  }
  return permissions;
};

export function getAuthorizationToken<T>(location: { query: { token: T } }) {
  return 'token' in location.query ? location.query.token : null;
}

export const getProvidersData = () => {
  const data = getScriptText('login-providers');
  try {
    return data && JSON.parse(data); // $flowfixme
  } catch (e) {
    return null;
  }
};

export const getPossibleErrorMessage = () => {
  const errorMessageElem = document.getElementById('errorMessage');
  const data = errorMessageElem && errorMessageElem.innerHTML;
  return data;
};

export function getSortedArrayByKey<KeyType>(arr: Array<{ [KeyType]: number }>, key: KeyType) {
  arr.sort((a, b) => {
    if (a[key] < b[key]) {
      return -1;
    } else if (a[key] > b[key]) {
      return 1;
    }
    return 0;
  });
  return arr;
}

export const isDateExpired = (date1: number, date2: number) => date1 > date2;

export const getNumberOfDays = (date1: number, date2: number) => {
  const days = (date1 - date2) / (1000 * 60 * 60 * 24);
  return Math.round(days);
};

export const calculatePercentage = (value1: number, value2: number) => Math.round(value1 * 100 / value2 * 100) / 100;

/*
  Handrolled instead of using lodash
  Because lodash/capitalize lowercases everything else
*/
export const capitalize = (s: string) => s.charAt(0).toUpperCase() + s.slice(1);

export const getDocumentScrollTop = () =>
  window.pageYOffset ||
  (document.documentElement && document.documentElement.scrollTop) ||
  (document.body && document.body.scrollTop) ||
  0;

export const getDomElementOffset = (el: HTMLElement) => {
  const rect = el.getBoundingClientRect();
  const scrollTop = getDocumentScrollTop();
  const scrollLeft =
    window.pageXOffset ||
    (document.documentElement && document.documentElement.scrollLeft) ||
    (document.body && document.body.scrollLeft) ||
    0;
  return { top: rect.top + scrollTop, left: rect.left + scrollLeft };
};

export const computeDomElementOffset = (ref: HTMLElement, offset: { top?: number, left?: number }) => {
  // inspired from jquery.setOffset
  const curOffset = getDomElementOffset(ref);
  const curCSS = window.getComputedStyle(ref);
  const curTop = parseFloat(curCSS.top) || 0;
  const curLeft = parseFloat(curCSS.left) || 0;
  const result = { top: curTop, left: curLeft };

  if (typeof offset.top === 'number') result.top = offset.top - curOffset.top + curTop;
  if (typeof offset.left === 'number') result.left = offset.left - curOffset.left + curLeft;
  return result;
};

export const createEvent = (
  typeArg: string,
  eventInit: { bubbles: boolean, cancelable: boolean } = { bubbles: false, cancelable: false }
) => {
  // inspired from https://developer.mozilla.org/en-US/docs/Web/API/Event/Event
  const event = document.createEvent('Event'); // we can't use 'new Event()' because ie
  event.initEvent(typeArg, eventInit.bubbles, eventInit.cancelable);
  return event;
};

/*
  Get basename from a unix or windows path
*/
export const getBasename = (path: string) =>
  path
    .split('\\')
    .pop()
    .split('/')
    .pop();

export const hashLinkScroll = () => {
  const { hash } = window.location;
  if (hash !== '') {
    const id = hash.replace('#', '').split('?')[0];
    // Push onto callback queue so it runs after the DOM is updated,
    // this is required when navigating from a different page so that
    // the element is rendered on the page before trying to getElementById.
    setTimeout(() => {
      const element = document.getElementById(id);
      if (element) {
        const offset = getDomElementOffset(element).top - 160;
        window.scrollTo({ top: offset, left: 0, behavior: 'smooth' });
      }
    }, 0);
  }
};

export const hexToRgb = (c: string) => {
  if (!c) return '';
  const hex = c.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255; // eslint-disable-line
  const g = (bigint >> 8) & 255; // eslint-disable-line
  const b = bigint & 255; // eslint-disable-line

  return [r, g, b].join();
};

export const isMobile = {
  android: () => navigator.userAgent.match(/Android/i),
  blackberry: () => navigator.userAgent.match(/BlackBerry/i),
  ios: () => navigator.userAgent.match(/iPhone|iPad|iPod/i),
  opera: () => navigator.userAgent.match(/Opera Mini/i),
  windows: () => navigator.userAgent.match(/IEMobile/i),
  any: () => isMobile.android() || isMobile.blackberry() || isMobile.ios() || isMobile.opera() || isMobile.windows()
};

// works for SCREAMING_SNAKE_CASE, snake_case or cRazY_SNAKE_case
export const snakeToCamel = (string: string) => string.toLowerCase().replace(/_[a-z]/g, match => match[1].toUpperCase());