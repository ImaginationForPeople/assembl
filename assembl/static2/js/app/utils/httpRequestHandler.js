import urljoin from 'url-join';
import Cookies from 'js-cookie';

import { getCSRFToken } from './csrf';
import { basePathV2 } from './server';

const convertToURLEncodedString = obj =>
  Object.keys(obj)
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(obj[k])}`)
    .join('&');
const getResponseContentType = xhr =>
  // eslint-disable-line no-unused-vars
  xhr.getResponseHeader('Content-Type').split(';')[0];

const useCSRFProtection = document.getElementById('useCSRFProtection') ? document.getElementById('useCSRFProtection').value : 'false';

/*
  A global async method that returns a Promisified ajax call
  @params payload [Object] The object that will be sent
  @params isJson [Boolean] Pass a flag if the object is JSON. Default is form header.
  @retuns [Promise]
*/

export const xmlHttpRequest = obj =>
  new Promise(async (resolve, reject) => {
    let payload = obj.payload;
    obj.headers = obj.headers || {}; // eslint-disable-line

    const xhr = new XMLHttpRequest();

    if (obj.method.toLowerCase() === 'post') {
      // Go and fetch a CSRF token for the POST request if activated
      if (useCSRFProtection === 'true') {
        await getCSRFToken();
        obj.headers['X-XSRF-TOKEN'] = Cookies.get('_csrf'); // eslint-disable-line
      }
    }

    if (obj.isJson && obj.isJson === true) {
      obj.headers['Content-Type'] = 'application/json'; // eslint-disable-line
      payload = JSON.stringify(obj.payload);
    } else {
      obj.headers['Content-Type'] = 'application/x-www-form-urlencoded'; // eslint-disable-line
      if (payload) {
        payload = convertToURLEncodedString(payload);
      }
    }

    const url = urljoin(basePathV2(obj), obj.url);
    xhr.open(obj.method, url);
    if (obj.headers) {
      Object.keys(obj.headers).forEach((key) => {
        xhr.setRequestHeader(key, obj.headers[key]);
      });
    }
    xhr.onload = () => {
      let resp;
      if (xhr.status >= 200 && xhr.status < 300) {
        resp = xhr.response;
        try {
          resp = JSON.parse(resp);
        } catch (e) {
          // TODO: Remove console warn AFTER a successful contract is agreed upon
          console.warn('A successful response did not return JSON. Passing status only.'); // eslint-disable-line
          resp = xhr.status;
        }
        resolve(resp);
      } else {
        // Contract agreed upon. If API is to fail, must respond with
        // JSONError type. Front-end respects this type of response only.
        const respType = getResponseContentType(xhr);
        if (respType === 'application/json') {
          resp = JSON.parse(xhr.responseText);
        } else {
          resp = [{ type: 'nonJson', message: '', status: xhr.status }];
        }
        reject(resp);
      }
    };
    xhr.onerror = () =>
      // Network level failure
      reject(xhr.statusText || xhr.responseText);
    xhr.send(payload);
  });

export const fetchContentType = url =>
  fetch(url, {
    method: 'HEAD'
  }).then(response => response.headers.get('Content-Type'), () => null);