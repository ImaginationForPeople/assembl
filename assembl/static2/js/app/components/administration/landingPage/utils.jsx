// @flow
import { browserHistory } from '../../../router';
import { get } from '../../../utils/routeMap';
import { getDiscussionSlug } from '../../../utils/globalFunctions';

export const goToModulesAdmin = () => {
  browserHistory.push(`${get('administration', { slug: getDiscussionSlug(), id: 'landingPage' })}`);
};