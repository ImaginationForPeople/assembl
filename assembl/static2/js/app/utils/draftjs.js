/*
  Utils related to draft-js library

  @flow
*/
import { type ContentBlock, type ContentState, EditorState, type EntityInstance } from 'draft-js';
import { convertFromHTML, convertToHTML } from 'draft-convert';
import { type List, type Map } from 'immutable';

// from our workspaces
/* eslint-disable import/no-extraneous-dependencies */
import { constants } from 'assembl-editor-utils';
import { converters as linkConverters } from 'draft-js-link-plugin';
import { converters as attachmentsConverters } from 'draft-js-attachment-plugin';
/* eslint-enable import/no-extraneous-dependencies */

import { addProtocol } from './linkify';
import { fromGlobalId } from './globalFunctions';

type Entry = {
  localeCode: string,
  value: string | EditorState
};

const { ENTITY_TYPES } = constants;

export function blockToHTML(block: ContentBlock): { start: string, end: string } | void {
  if (block.type === 'atomic') {
    return { start: '<div class="atomic-block" data-blocktype="atomic">', end: '</div>' };
  }

  return undefined;
}

export function htmlToBlock(nodeName: string, node: HTMLElement, lastList: *, inBlock: string): void | string {
  const isAtomicBlock = nodeName === 'div' && node.dataset.blocktype === 'atomic';
  if (isAtomicBlock || (nodeName === 'img' && inBlock !== 'atomic')) {
    return 'atomic';
  }

  return undefined;
}

const customConvertFromHTML = convertFromHTML({
  htmlToBlock: htmlToBlock,
  htmlToEntity: function (nodeName: string, node: HTMLElement, createEntity: Function): EntityInstance | null {
    if (nodeName === 'a') {
      // $FlowFixMe: if nodeName is 'a', node should be an HTMLAnchorElement
      return linkConverters.htmlToEntity(nodeName, node, createEntity, addProtocol);
    }

    return attachmentsConverters.htmlToEntity(nodeName, node, createEntity);
  }
});

const customConvertToHTML = convertToHTML({
  blockToHTML: blockToHTML,
  entityToHTML: (entity: EntityInstance, originalText: string): string => {
    if (entity.type === ENTITY_TYPES.document || entity.type === ENTITY_TYPES.image) {
      return attachmentsConverters.entityToHTML(entity);
    } else if (entity.type === ENTITY_TYPES.link) {
      return linkConverters.entityToHTML(entity, originalText);
    }

    return originalText;
  }
});

export function convertEntries(converter: Function): Function {
  return function (entries: Array<Entry>): Array<Entry> {
    return entries.map(entry => ({
      ...entry,
      value: converter(entry.value)
    }));
  };
}

export function convertToEditorState(value: string): EditorState {
  if (value) {
    return EditorState.createWithContent(customConvertFromHTML(value));
  }

  return EditorState.createEmpty();
}

export const convertEntriesToEditorState = convertEntries(convertToEditorState);

export const convertContentStateToHTML = (cs: ContentState): string => customConvertToHTML(cs);

export const convertEditorStateToHTML = (es: EditorState): string => convertContentStateToHTML(es.getCurrentContent());

export function convertImmutableEntriesToJS(entries: List<Map<string, any>>): Array<Object> {
  return entries.map(entry => entry.toObject()).toArray();
}

export const convertEntriesToHTML = convertEntries(convertEditorStateToHTML);

export function editorStateIsEmpty(editorState: EditorState): boolean {
  const contentState = editorState.getCurrentContent();
  const containAtomicBlock = blockMap => blockMap.some(b => b.type === 'atomic');
  if (contentState.getPlainText().length === 0 && !containAtomicBlock(contentState.getBlockMap())) {
    return true;
  }

  return false;
}

export type UploadNewAttachmentsPromiseResult = {
  contentState: ContentState,
  documentIds: Array<string>
};

export function uploadNewAttachments(
  editorState: EditorState,
  uploadDocument: Function
): Promise<UploadNewAttachmentsPromiseResult> {
  const documentIds = [];
  let contentState = editorState.getCurrentContent();
  const entities = [];
  contentState.getBlockMap().forEach((block) => {
    if (block.type === 'atomic') {
      block.findEntityRanges((entityRange) => {
        const entityKey = entityRange.entity;
        if (entityKey) {
          const entity = contentState.getEntity(entityKey);
          if (entity) {
            entities.push({
              entityKey: entityKey,
              entity: entity
            });
          }
        }
      });
    }
  });

  let uploadDocumentsPromise = Promise.resolve();
  entities.forEach((entityInfo) => {
    const { entity, entityKey } = entityInfo;
    if (entity.data.src instanceof File) {
      // this is a new document, add a promise to create it and modify its entity
      const variables = {
        file: entity.data.src
      };
      uploadDocumentsPromise = uploadDocumentsPromise.then(() =>
        uploadDocument({ variables: variables }).then((res) => {
          if (res && res.data) {
            const doc = res.data.uploadDocument.document;
            const docId = fromGlobalId(doc.id);
            documentIds.push(docId);
            // update entity
            const { externalUrl, mimeType, title } = doc;
            contentState = contentState.replaceEntityData(entityKey, {
              id: docId,
              src: externalUrl,
              title: title,
              mimeType: mimeType
            });
          }
        })
      );
    } else {
      documentIds.push(entity.data.id);
    }
  });

  return uploadDocumentsPromise.then(
    () =>
      new Promise((resolve) => {
        resolve({
          contentState: contentState,
          // $FlowFixMe
          documentIds: documentIds
        });
      })
  );
}