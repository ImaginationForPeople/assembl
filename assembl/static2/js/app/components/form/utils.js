// @flow
import { I18n } from 'react-redux-i18n';

import type { I18nValue, FileValue, FileVariable, MutationsPromises, SaveStatus } from './types.flow';
import { displayAlert } from '../../utils/utilityManager';
import { runSerial } from '../administration/saveButton';

export function i18nValueIsEmpty(v: I18nValue): boolean {
  return (
    !v ||
    Object.keys(v)
      .map(key => v[key]) // flow doesn't treat Object.values as expected, see: https://github.com/facebook/flow/issues/2221
      .every(s => s.length === 0)
  );
}

export function convertEntries(_entries: LangstringEntries): I18nValue {
  const entries = _entries || [];
  return entries.reduce((result, item) => {
    const { localeCode, value } = item;
    return {
      ...result,
      [localeCode]: value
    };
  }, {});
}

export function getValidationState(error: ?string, touched: ?boolean): ?string {
  return touched && error ? 'error' : null;
}

export const createSave = (successMsgId: string) => async (mutationsPromises: MutationsPromises): Promise<SaveStatus> => {
  let status = 'PENDING';
  await runSerial(mutationsPromises)
    .then(() => {
      status = 'OK';
      displayAlert('success', I18n.t(successMsgId));
    })
    .catch((error) => {
      status = 'KO';
      displayAlert('danger', error.message, false, 30000);
    });

  return status;
};

export function convertToEntries(valuesByLocale: I18nValue): LangstringEntries {
  return Object.keys(valuesByLocale).map(locale => ({
    localeCode: locale,
    value: valuesByLocale[locale]
  }));
}

export function getFileVariable(img: FileValue, initialImg: FileValue): FileVariable {
  if (initialImg && !img) {
    return 'TO_DELETE';
  }

  // If thematic.img.externalUrl is an object, it means it's a File.
  // We need to send image: null if we didn't change the image.
  const variab = img && typeof img.externalUrl === 'object' ? img.externalUrl : null;
  return variab;
}