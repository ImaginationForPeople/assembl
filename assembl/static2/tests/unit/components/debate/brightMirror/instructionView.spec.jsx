import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { connectedUserCan } from '../../../../../js/app/utils/permissions';

import InstructionView from '../../../../../js/app/components/debate/brightMirror/InstructionView';

jest.mock('../../../../../js/app/utils/permissions');

describe('Bright Mirror page', () => {
  it('should match BrightMirror instruction snapshot', () => {
    connectedUserCan.mockImplementation(() => true);

    const timeline = [
      {
        identifier: 'brightMirror',
        start: '1900-01-01T02:00:00Z',
        end: '2200-01-01T00:00:00Z',
        title: { entries: [{ en: 'brightMirror' }] }
      }
    ];
    const props = {
      announcementContent: {
        title: 'Instruction',
        body: 'Participez à la création de fiction'
      },
      isUserConnected: true,
      ideaId: 0,
      refetchIdea: Function,
      posts: [],
      timeline: timeline,
      identifier: 'brightMirror'
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<InstructionView {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('timeline is over', () => {
    const timeline = [
      {
        identifier: 'brightMirror',
        start: '1900-01-01T02:00:00Z',
        end: '2000-01-01T00:00:00Z',
        title: { entries: [{ en: 'brightMirror' }] }
      }
    ];
    const props = {
      announcementContent: {
        title: 'Instruction',
        body: 'Participez à la création de fiction'
      },
      isUserConnected: true,
      ideaId: 0,
      refetchIdea: Function,
      posts: [],
      timeline: timeline,
      identifier: 'brightMirror'
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<InstructionView {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('user cannot post', () => {
    connectedUserCan.mockImplementation(() => false);

    const timeline = [
      {
        identifier: 'brightMirror',
        start: '1900-01-01T02:00:00Z',
        end: '2200-01-01T00:00:00Z',
        title: { entries: [{ en: 'brightMirror' }] }
      }
    ];
    const props = {
      announcementContent: {
        title: 'Instruction',
        body: 'Participez à la création de fiction'
      },
      isUserConnected: true,
      ideaId: 0,
      refetchIdea: Function,
      posts: [],
      timeline: timeline,
      identifier: 'brightMirror'
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<InstructionView {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('user is not connected', () => {
    const timeline = [
      {
        identifier: 'brightMirror',
        start: '1900-01-01T02:00:00Z',
        end: '2000-01-01T00:00:00Z',
        title: { entries: [{ en: 'brightMirror' }] }
      }
    ];
    const props = {
      announcementContent: {
        title: 'Instruction',
        body: 'Participez à la création de fiction'
      },
      isUserConnected: false,
      ideaId: 0,
      refetchIdea: Function,
      posts: [],
      timeline: timeline,
      identifier: 'brightMirror'
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<InstructionView {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});