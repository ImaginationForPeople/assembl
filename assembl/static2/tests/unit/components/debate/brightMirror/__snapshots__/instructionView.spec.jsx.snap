// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bright Mirror page should hide fiction creation section if timeline is over 1`] = `
<div
  className="instruction-view"
>
  <InstructionsText
    body="Participez à la création de fiction"
    title="Instruction"
  />
  <div
    className="overflow-x"
  >
    <FictionsList
      identifier="brightMirror"
      lang={undefined}
      posts={Array []}
    />
  </div>
</div>
`;

exports[`Bright Mirror page should hide fiction creation section if user is not allowed post 1`] = `
<div
  className="instruction-view"
>
  <InstructionsText
    body="Participez à la création de fiction"
    title="Instruction"
  />
  <div
    className="overflow-x"
  >
    <FictionsList
      identifier="brightMirror"
      lang={undefined}
      posts={Array []}
    />
  </div>
</div>
`;

exports[`Bright Mirror page should hide fiction creation section if user is not connected 1`] = `
<div
  className="instruction-view"
>
  <InstructionsText
    body="Participez à la création de fiction"
    title="Instruction"
  />
  <div
    className="overflow-x"
  >
    <FictionsList
      identifier="brightMirror"
      lang={undefined}
      posts={Array []}
    />
  </div>
</div>
`;

exports[`Bright Mirror page should match BrightMirror instruction snapshot 1`] = `
<div
  className="instruction-view"
>
  <InstructionsText
    body="Participez à la création de fiction"
    title="Instruction"
  />
  <div
    className="overflow-x"
  >
    <TopPostFormContainer
      bodyPlaceholder="debate.brightMirror.fillBodyLabel"
      fillBodyLabel="debate.brightMirror.fillBodyLabel"
      ideaId={0}
      instructionLabel="debate.brightMirror.startFictionLabel"
      postSuccessMsg="debate.brightMirror.postSuccessMsg"
      refetchIdea={[Function]}
      topPostsCount={0}
    />
    <FictionsList
      identifier="brightMirror"
      lang={undefined}
      posts={Array []}
    />
  </div>
</div>
`;
